botocore/client.py

```diff
@@ -325,6 +246,144 @@ def _api_call(self, *args, **kwargs):
         return _api_call
 
 
+class ClientEndpointBridge(object):
+    """Bridges endpoint data and client creation
+
+    This class handles taking out the relevant arguments from the endpoint
+    resolver and determining which values to use, taking into account any
+    client configuration options and scope configuration options.
+
+    This class also handles determining what, if any, region to use if no
+    explicit region setting is provided. For example, Amazon S3 client will
+    utilize "us-east-1" by default if no region can be resolved."""
+
+    SIGNATURE_PRECEDENCE = ['s3', 's3v4', 'v4', 'v2']
```

Comment:
> It's both a future proof thing and an easy way to define which signers we support in Botocore. If we find one of the preferred signers in the list of signature versions, then we know that we can send requests to a service.

That's handled from the auth type maps though isn't it? https://github.com/boto/botocore/blob/develop/botocore/auth.py#L736

>The reason to check for v4 and prefer it over v2 is to correctly handle the case of one of the few remaining v2 services adding support for v4.  At that point, it is unlikely the service would stop supporting v2, so we'd need to account for the case of both v2 and v4 as available signature versions.

I'm comparing how we've rolled out v4 to services so far.  With the exception of S3 (which has its own special signature version anyways) the `v2->v4` upgrade has been completely transparent.  We just update the signatureVersion to say `v4` in the service model.  I picked a handful of services and flipped them back to v2, and they all still work (EC2, ELB, IAM, etc.).  I suspect every service that once supported v2 still supports v2.  Yet we don't call them out here in this endpoints file.  And I don't think we need to.  If the last remaining v2 service, importexport, flips to v4, I imagine we'd just update this file to say that we should start using `v4` for importexport wouldn't we?  There's no need to specify both signature versions in this file.

I'm just thinking if maybe this is just yet-another special case required for S3 only.